File: tests/unit/task_contract_golden_test.go
package unit

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/madebyduy/RhinoQ/pkg/rhinoq"
)

type taskContractV1 struct {
	Snapshot rhinoq.TaskSnapshot `json:"snapshot"`
	Result   rhinoq.TaskResult   `json:"result"`
}

func TestPublicTaskContractMatchesGoldenV1(t *testing.T) {
	t.Parallel()
	total := int64(10)
	updatedAt := time.Date(2026, 7, 29, 14, 2, 0, 0, time.UTC)
	contract := taskContractV1{
		Snapshot: rhinoq.TaskSnapshot{
			SchemaVersion: 1,
			EntityVersion: 7,
			ID:            "task-contract-01",
			Type:          "report.export",
			State:         rhinoq.TaskRunning,
			Progress: rhinoq.TaskProgress{
				Completed: 4,
				Total:     &total,
				Message:   "rendering",
			},
			HasResult: true,
			Executions: []rhinoq.TaskExecutionSummary{{
				ID:      "execution-contract-01",
				Attempt: 1,
				Runtime: "bullmq",
				State:   "running",
				Version: 3,
			}},
			CreatedAt: time.Date(2026, 7, 29, 14, 0, 0, 0, time.UTC),
			UpdatedAt: updatedAt,
		},
		Result: rhinoq.TaskResult{
			SchemaVersion: 1,
			EntityVersion: 7,
			TaskID:        "task-contract-01",
			Reference:     "s3://reports/task-contract-01.pdf",
			UpdatedAt:     updatedAt,
		},
	}
	actual, err := json.MarshalIndent(contract, "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	actual = append(actual, '\n')
	goldenPath := filepath.Join(
		repositoryRoot(t), "testdata", "contracts", "task-contract-v1.json",
	)
	expected, err := os.ReadFile(goldenPath)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(actual, expected) {
		t.Fatalf(
			"Task wire contract drifted; update schema/version and both SDKs before changing %s\nactual:\n%s",
			goldenPath,
			actual,
		)
	}
}

File: sdks/node/test/task-contract-golden.test.mjs
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import test from 'node:test';

import { RhinoQClient } from '../dist/index.js';

const goldenURL = new URL(
  '../../../testdata/contracts/task-contract-v1.json',
  import.meta.url,
);

test('Node Task client consumes the shared Go wire contract v1', async () => {
  const golden = JSON.parse(await readFile(goldenURL, 'utf8'));
  const client = new RhinoQClient({
    url: 'http://gateway.test',
    fetch: async (url) => {
      if (url.endsWith('/result')) {
        return Response.json(golden.result);
      }
      return Response.json(golden.snapshot);
    },
  });

  const snapshot = await client.getTask('task-contract-01');
  const result = await client.getTaskResult('task-contract-01');

  assert.deepEqual(snapshot, golden.snapshot);
  assert.deepEqual(result, golden.result);
  assert.deepEqual(Object.keys(snapshot), [
    'schemaVersion',
    'entityVersion',
    'id',
    'type',
    'ownerId',
    'state',
    'cancellation',
    'progress',
    'hasResult',
    'executions',
    'createdAt',
    'updatedAt',
  ]);
  assert.deepEqual(Object.keys(snapshot.executions[0]), [
    'id',
    'attempt',
    'runtime',
    'state',
    'version',
  ]);
});
